src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java

```diff
@@ -0,0 +1,472 @@
+                                Read record = pair.first();
+                                record.setDuplicateFragment(true);
+                                context.output(record);
+
+                                record = pair.second();
+                                record.setDuplicateFragment(true);
+                                context.output(record);
+                            }
+                        }
+                    });
+        }
+
+
+        /**
+         * Takes the reads,
+         * group them by library, contig, position and orientation,
+         * within each group
+         *   (a) if there are only fragments, mark all but the highest scoring as duplicates, or,
+         *   (b) if at least one is marked as paired, mark all fragments as duplicates.
+         *  Note: Emit only the fragments, as the paired reads are handled separately.
+         */
+        PCollection<Read> transformFragments(final PCollectionView<SAMFileHeader> headerPcolView, final PCollection<Read> fragments) {
+            final PCollection<KV<String, Iterable<Read>>> readsGroupedByKey = groupReadsByKey(fragments, headerPcolView);
+            final PTransform<PCollection<? extends KV<String, Iterable<Read>>>, PCollection<Read>> markGroupedDuplicateFragments = markGroupedDuplicateFragments();
+            return readsGroupedByKey
+                    .apply(markGroupedDuplicateFragments);//no need to set up coder for Read (uses GenericJsonCoder)
+        }
+
+        private PTransform<PCollection<? extends KV<String, Iterable<Read>>>, PCollection<Read>> markGroupedDuplicateFragments() {
+            return ParDo.named("mark dups")
+                    .of(new DoFn<KV<String, Iterable<Read>>, Read>() {
+                        private static final long serialVersionUID = 1l;
+
+                        @Override
+                        public void processElement(final ProcessContext context) throws Exception {
+                            final Map<Boolean, List<Read>> byPairing = StreamSupport.stream(context.element().getValue().spliterator(), false).collect(Collectors.partitioningBy(
+                                    read -> GenomicsReadUtils.isPaired(read)
+                            ));
+                            //
+                            // The existence of any paired end means we mark all fragments as duplicates.
+                            // Otherwise, mark all but the highest scoring fragment.
+                            // Note the we emit only fragments from this mapper.
+                            if (byPairing.get(true).isEmpty()) {
+                                final List<Read> frags = Ordering.natural().reverse().onResultOf((Read read) -> score(read)).immutableSortedCopy(byPairing.get(false));
+                                if (!frags.isEmpty()) {
+                                    context.output(Iterables.getFirst(frags, null));
```

Comment:
It looks like you might output null (but the line above indicates you wouldn't, right?) Why not just do frags.get(0) since !frags.isEmpty().